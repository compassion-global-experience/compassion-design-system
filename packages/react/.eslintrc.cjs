module.exports = {
  extends: [
    'standard-jsx',
    'standard-react',
    'plugin:jsx-a11y/recommended',
    'plugin:react/recommended',
    'plugin:react/jsx-runtime',
    'plugin:react-hooks/recommended',
    'plugin:storybook/recommended',
  ],
  env: {
    browser: true,
    es6: true,
    jest: true,
    node: false,
  },
  rules: {
    'jsx-quotes': ['error', 'prefer-double'],
  },
};
