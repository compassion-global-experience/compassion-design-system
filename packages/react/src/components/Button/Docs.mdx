import { Markdown } from '@storybook/blocks';
import {
  Canvas,
  Controls,
  Meta,
  Stories,
  Story,
  Title,
} from '@storybook/addon-docs';

import * as ButtonStories from './Button.stories.tsx';

import _01_Overview from '@compassion-design-system/core/src/components/Button/docs/01_Overview.md';
import _02_Usage from '@compassion-design-system/core/src/components/Button/docs/02_Usage.md';
import _03_Kind from '@compassion-design-system/core/src/components/Button/docs/03_Kind.md';
import _04_Emphasis from '@compassion-design-system/core/src/components/Button/docs/04_Emphasis.md';
import _05_Size from '@compassion-design-system/core/src/components/Button/docs/05_Size.md';
import _06_State from '@compassion-design-system/core/src/components/Button/docs/06_State.md';
import _07_OtherProperties from '@compassion-design-system/core/src/components/Button/docs/07_OtherProperties.md';
import _08_BestPractices from '@compassion-design-system/core/src/components/Button/docs/08_BestPractices.md';

<Meta of={ButtonStories} />

<Title />

<Markdown>{_01_Overview}</Markdown>

<Canvas of={ButtonStories.Default} sourceState="shown" withToolbar="true" />
<Controls of={ButtonStories.Default} />

<Markdown>{_02_Usage}</Markdown>

## Component Properties

<Markdown>{_03_Kind}</Markdown>
<Canvas of={ButtonStories.Kinds} />

<Markdown>{_04_Emphasis}</Markdown>
<Canvas of={ButtonStories.Emphasis} />

<Markdown>{_05_Size}</Markdown>
<Canvas of={ButtonStories.Size} />

<Markdown>{_06_State}</Markdown>
<Canvas of={ButtonStories.State} />

<Markdown>{_07_OtherProperties}</Markdown>
<Canvas of={ButtonStories.WithIcon} />

<Markdown>{_08_BestPractices}</Markdown>
