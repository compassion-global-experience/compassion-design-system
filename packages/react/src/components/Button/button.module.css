@import "~@compassion-gds/css/src/components/Button/button.module.css";
