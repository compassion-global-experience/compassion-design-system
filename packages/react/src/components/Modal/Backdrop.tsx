const Backdrop = () =>
  <span style={{ position: 'fixed', inset: 0, background: 'black', opacity: 0.7 }} />;

export default Backdrop;
