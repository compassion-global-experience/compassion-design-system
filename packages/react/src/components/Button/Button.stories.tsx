import { Meta } from '@storybook/react';
import Button from './Button';
import { ArrowLeft, ArrowRight, Info } from 'phosphor-react';

const meta: Meta<typeof Button> = {
  /* ðŸ‘‡ The title prop is optional.
   * See https://storybook.js.org/docs/7.0/react/configure/overview#configure-story-loading
   * to learn how to generate automatic titles
   */
  title: 'Components/Button',
  component: Button,
};

export default meta;

// More on args: https://storybook.js.org/docs/react/writing-stories/args
export const Primary = {
  args: {
    mode: 'primary',
    label: 'Button',
  },
};

export const Secondary = {
  args: {
    mode: 'secondary',
    label: 'Button',
  },
};

export const Tertiary = {
  args: {
    mode: 'tertiary',
    label: 'Button',
  },
};

export const Small = {
  args: {
    size: 'small',
    label: 'Button',
  },
};

export const Medium = {
  args: {
    size: 'medium',
    label: 'Button',
  },
};

export const Large = {
  args: {
    size: 'large',
    label: 'Button',
  },
};

export const Disabled = {
  args: {
    disabled: true,
    label: 'Button',
  },
};

export const WithIconRight: typeof meta = {
  args: {
    mode: 'primary',
  },
  render: (args) => (
    <Button {...args}>
      Log in <ArrowRight />
    </Button>
  ),
};

export const WithIconLeft: typeof meta = {
  args: {
    mode: 'primary',
  },
  render: (args) => (
    <Button {...args}>
      <ArrowLeft /> Go back
    </Button>
  ),
};

export const WithIconOnBothSides: typeof meta = {
  args: {
    mode: 'primary',
  },
  render: (args) => (
    <Button {...args}>
      <ArrowLeft /> Options <ArrowRight />
    </Button>
  ),
};

export const IconOnly = {
  args: {
    mode: 'primary',
    size: 'small',
    label: <Info size={24} />,
  },
};
