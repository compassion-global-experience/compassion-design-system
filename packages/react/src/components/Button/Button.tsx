'use client';

import { MouseEvent, ButtonHTMLAttributes, forwardRef } from 'react';
import styles from '@compassion-design-system/core/src/components/Button/button.module.css';
import { getClasses } from '../../utils/classes';
import { cdsIconMap } from '../Icon/Icon';

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  label: string;
  showLabel?: boolean;
  kind?: 'cta' | 'white' | 'inverted' | 'destructive';
  emphasis?: 'primary' | 'secondary' | 'tertiary';
  size?: 'small' | 'medium' | 'large';
  disabled?: boolean;
  icon?: keyof typeof cdsIconMap;
  iconDescription?: string;
  iconPosition?: 'start' | 'end';
  onClick?: (event: MouseEvent<HTMLButtonElement>) => void;
}

const Button = forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => {
  const {
    label,
    showLabel = true,
    kind,
    emphasis = 'primary',
    size = 'medium',
    disabled = false,
    className,
    style,
    onClick,
    icon,
    iconDescription,
    iconPosition,
    ...rest
  } = props;

  const IconComponent = cdsIconMap[icon];

  const buttonSizeToIconSizeMap: { [key: string]: number } = {
    small: 13,
    medium: 16,
    large: 20,
  };

  const ButtonIcon = () => {
    return (
      <div className={getClasses(styles, buttonIconClassList)}>
        <IconComponent
          className=""
          alt={iconDescription}
          size={buttonSizeToIconSizeMap[size]}
        />
      </div>
    );
  };

  const buttonClassList = [`cds-button--${emphasis}`, `cds-button--${size}`];

  if (kind) {
    buttonClassList.push(`cds-button--${kind}`);
  }

  const buttonIconClassList = ['cds-button__icon'];

  if (IconComponent && showLabel) {
    if (iconPosition && iconPosition === 'end') {
      buttonClassList.push('cds-button--reverseTextIcon');
      buttonIconClassList.push('cds-button__icon--end');
    } else {
      buttonIconClassList.push('cds-button__icon--start');
    }
  }

  const buttonClassNames = getClasses(styles, buttonClassList, className);

  return (
    <button
      onClick={onClick}
      className={buttonClassNames}
      disabled={disabled}
      style={style}
      ref={ref}
      {...rest}
    >
      {IconComponent && <ButtonIcon />}
      {showLabel && label}
    </button>
  );
});

Button.displayName = 'Button';

export default Button;
