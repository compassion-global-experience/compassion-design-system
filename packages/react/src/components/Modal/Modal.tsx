import { CSSProperties, MouseEvent, ReactNode } from 'react';
import '@compassion-gds/css/src/components/Modal/modal.css';
import { X } from '../icons';

export interface ModalProps {
  title?: ReactNode,
  size?: 'default' | 'small' | 'wide',
  footerSlot?: ReactNode,
  children: ReactNode,
  onClose?: () => void,
  className?: string,
  style?: CSSProperties,
}

const Modal = ({ title = '', size = 'default', onClose, footerSlot, className, style, children }: ModalProps) => {
    const classNames = ['modal', size, className].join(' ');

    return (
      <div className={classNames} style={style}>
        {onClose && <ModalClose onClick={onClose} />}
        {title && <div className="modal-header">{title}</div>}
        <div className="modal-body">{children}</div>
        {Boolean(footerSlot) && <div className="modal-footer">{footerSlot}</div>}
      </div>
  );
};

export interface ButtonProps {
  onClick: (event: MouseEvent<HTMLButtonElement>) => void,
}

const ModalClose = ({ onClick }: ButtonProps) => (
  <button type="button" className="modal-button-close" onClick={onClick}>
    <X size={24} />
  </button>
);

export default Modal;
