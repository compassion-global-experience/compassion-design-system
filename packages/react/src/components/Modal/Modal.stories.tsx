import { useState } from 'react';
import { Meta } from '@storybook/react';
import Modal from './Modal';
import Backdrop from './Backdrop';
import Button from '../Button';

const meta: Meta<typeof Modal> = {
  title: 'Components/Modal',
  component: Modal,
};

export default meta;

const Template = props => {
  const [open, setOpen] = useState(false);

  return (
    <>
      {open && <Backdrop />}
      <button type="button" className="button-main secondary medium" onClick={() => setOpen(true)}>
        Open Modal
      </button>
      {
          open &&
            <Modal {...props} onClose={() => setOpen(false)}>
              <div style={{ border: '3px dashed', textAlign: 'center', borderRadius: 4, margin: 10, padding: 10 }}>
                Slot
              </div>
            </Modal>
      }
    </>
  );
};

const ModalFooter = () => (
  <>
    <Button label="Button 2" mode="secondary" />
    <Button label="Button 1" mode="primary" />
  </>
);

export const Default = props => <Template {...props} />;
Default.args = {
  title: 'Title',
  size: 'default',
  footerSlot: <ModalFooter />,
};

export const Small = props => <Template {...props} />;
Small.args = {
  title: 'Title',
  size: 'small',
  footerSlot: <ModalFooter />,
};

export const Wide = props => <Template {...props} />;
Wide.args = {
  title: 'Title',
  size: 'wide',
  footerSlot: <ModalFooter />,
};

export const NoHeader = props => <Template {...props} />;
NoHeader.args = {
  footerSlot: <ModalFooter />,
};

export const NoFooter = props => <Template {...props} />;
NoFooter.args = {
  title: 'Title',
};

export const NoHeaderAndFooter = props => <Template {...props} />;
NoHeaderAndFooter.args = {};

