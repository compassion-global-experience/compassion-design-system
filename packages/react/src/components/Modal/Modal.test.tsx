import { render } from '@testing-library/react';

import Modal from './Modal';
import Button from '../Button';

describe('Modal', () => {
  it('should render `Default` size with `Header` and `Footer`', () => {
    const { container } = renderComponent({ size: 'default' });
    expect(container).toMatchSnapshot();
  });

  it('should render `Wide` size without `Header`', () => {
    const { container } = renderComponent({ size: 'wide', title: undefined });
    expect(container).toMatchSnapshot();
  });

  it('should render `Small` size without `Footer`', () => {
    const { container } = renderComponent({
      size: 'small',
      footerSlot: undefined,
    });
    expect(container).toMatchSnapshot();
  });

  it('should render `Default` size without `Header` and `Footer`', () => {
    const { container } = renderComponent({
      size: 'small',
      footerSlot: undefined,
    });
    expect(container).toMatchSnapshot();
  });

  const renderComponent = (props) =>
    render(
      <Modal
        title="Title"
        footerSlot={<ModalFooter />}
        onClose={() => undefined}
        {...props}
      >
        <div>Slot</div>
      </Modal>,
    );
});

const ModalFooter = () => (
  <>
    <Button label="Button 2" emphasis="secondary" />
    <Button label="Button 1" emphasis="primary" />
  </>
);
