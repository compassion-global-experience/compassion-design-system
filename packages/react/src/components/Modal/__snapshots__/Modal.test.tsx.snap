// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render \`Default\` size with \`Header\` and \`Footer\` 1`] = `
<div>
  <div
    class="cds-modal cds-modal--default"
  >
    <button
      class="cds-modal__button--close"
      type="button"
    >
      <svg
        fill="currentColor"
        height="24"
        viewBox="0 0 256 256"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M205.66,194.34a8,8,0,0,1-11.32,11.32L128,139.31,61.66,205.66a8,8,0,0,1-11.32-11.32L116.69,128,50.34,61.66A8,8,0,0,1,61.66,50.34L128,116.69l66.34-66.35a8,8,0,0,1,11.32,11.32L139.31,128Z"
        />
      </svg>
    </button>
    <div
      class="cds-modal__header"
    >
      Title
    </div>
    <div
      class="cds-modal__body"
    >
      <div>
        Slot
      </div>
    </div>
    <div
      class="cds-modal__footer"
    >
      <button
        class="cds-button--secondary cds-button--medium"
      >
        Button 2
      </button>
      <button
        class="cds-button--primary cds-button--medium"
      >
        Button 1
      </button>
    </div>
  </div>
</div>
`;

exports[`Modal should render \`Default\` size without \`Header\` and \`Footer\` 1`] = `
<div>
  <div
    class="cds-modal cds-modal--small"
  >
    <button
      class="cds-modal__button--close"
      type="button"
    >
      <svg
        fill="currentColor"
        height="24"
        viewBox="0 0 256 256"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M205.66,194.34a8,8,0,0,1-11.32,11.32L128,139.31,61.66,205.66a8,8,0,0,1-11.32-11.32L116.69,128,50.34,61.66A8,8,0,0,1,61.66,50.34L128,116.69l66.34-66.35a8,8,0,0,1,11.32,11.32L139.31,128Z"
        />
      </svg>
    </button>
    <div
      class="cds-modal__header"
    >
      Title
    </div>
    <div
      class="cds-modal__body"
    >
      <div>
        Slot
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal should render \`Small\` size without \`Footer\` 1`] = `
<div>
  <div
    class="cds-modal cds-modal--small"
  >
    <button
      class="cds-modal__button--close"
      type="button"
    >
      <svg
        fill="currentColor"
        height="24"
        viewBox="0 0 256 256"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M205.66,194.34a8,8,0,0,1-11.32,11.32L128,139.31,61.66,205.66a8,8,0,0,1-11.32-11.32L116.69,128,50.34,61.66A8,8,0,0,1,61.66,50.34L128,116.69l66.34-66.35a8,8,0,0,1,11.32,11.32L139.31,128Z"
        />
      </svg>
    </button>
    <div
      class="cds-modal__header"
    >
      Title
    </div>
    <div
      class="cds-modal__body"
    >
      <div>
        Slot
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal should render \`Wide\` size without \`Header\` 1`] = `
<div>
  <div
    class="cds-modal cds-modal--wide"
  >
    <button
      class="cds-modal__button--close"
      type="button"
    >
      <svg
        fill="currentColor"
        height="24"
        viewBox="0 0 256 256"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M205.66,194.34a8,8,0,0,1-11.32,11.32L128,139.31,61.66,205.66a8,8,0,0,1-11.32-11.32L116.69,128,50.34,61.66A8,8,0,0,1,61.66,50.34L128,116.69l66.34-66.35a8,8,0,0,1,11.32,11.32L139.31,128Z"
        />
      </svg>
    </button>
    <div
      class="cds-modal__body"
    >
      <div>
        Slot
      </div>
    </div>
    <div
      class="cds-modal__footer"
    >
      <button
        class="cds-button--secondary cds-button--medium"
      >
        Button 2
      </button>
      <button
        class="cds-button--primary cds-button--medium"
      >
        Button 1
      </button>
    </div>
  </div>
</div>
`;
